### DISCUSSION TITLE
Relations and foreign keys

### DISCUSSION THREAD

**Discussion by Ahmadh26** [prisma/prisma#12731] [Q&A]
Hey, so i'm writing up my `schema.prisma` and i'm facing a problem with foreign keys and relations

i have the following: 

```ts
model User {
  id            String    @id @default(uuid())
  name          String?
  email         String?   @unique
  emailVerified DateTime?
  nsid          String?
  image         String?
  type          user_type @default(default_user)
  accounts      Account[]
  sessions      Session[]
  author        Author?   @relation(fields: [authorId], references: [id], onDelete: SetNull, onUpdate: Cascade)
  authorId      String?
  location      String?
  createdAt     DateTime  @default(now())
}

model Author {
  id             String                @id @default(uuid())
  display_name   String
  web_url        String?
  details        String?               @db.Text
  logo           String?
  status         author_status
  exclusive      Boolean
  type           author_type
  message        String?               @db.Text
  work_type      author_work_type
  accepted_terms author_accepted_terms
  fee_id         Int
  external_url   String?               @db.Text
  createdAt      DateTime              @default(now())
  user           User[]
}
```

whenever a user is created, logged in using a 3rd party application, i get the following error: 

```bash
  3 exports.PrismaAdapter = void 0;
  4 function PrismaAdapter(p) {
  5     return {
→ 6         createUser: (data) => p.user.create(
  Foreign key constraint failed on the field: `authorId` {
  message: '\n' +
    'Invalid `p.user.create()` invocation in\n' +
    'C:\\Users\\a_4n\\Desktop\\repos\\ns-v2\\node_modules\\@next-auth\\prisma-adapter\\dist\\index.js:6:38\n' +
    '\n' +
    '  3 exports.PrismaAdapter = void 0;\n' +
    '  4 function PrismaAdapter(p) {\n' +
    '  5     return {\n' +
    '→ 6         createUser: (data) => p.user.create(\n' +
    '  Foreign key constraint failed on the field: `authorId`',
  stack: 'Error: \n' +
    'Invalid `p.user.create()` invocation in\n' +
    'C:\\Users\\a_4n\\Desktop\\repos\\ns-v2\\node_modules\\@next-auth\\prisma-adapter\\dist\\index.js:6:38\n' +
    '\n' +
    '  3 exports.PrismaAdapter = void 0;\n' +
    '  4 function PrismaAdapter(p) {\n' +
    '  5     return {\n' +
    '→ 6         createUser: (data) => p.user.create(\n' +
    '  Foreign key constraint failed on the field: `authorId`\n' +
    '    at Object.request (C:\\Users\\a_4n\\Desktop\\repos\\ns-v2\\node_modules\\@prisma\\client\\runtime\\index.js:45578:15)\n' +
    '    at async PrismaClient._request (C:\\Users\\a_4n\\Desktop\\repos\\ns-v2\\node_modules\\@prisma\\client\\runtime\\index.js:46405:18)',
  name: 'Error'
}
```

I do understand what the error means, but i can't figure out how to make the authorId accepted as null if the user doesn't belong to an author.

any help is appreciated.

**Comment by janpio** [comment]
What database are you trying to use?
What data are you trying to write here when this error appears?

**Reply by Ahmadh26** [reply]
I am using MySQL

the data is being written is the following in the user table (since i am doing an OAuth signin with github): 
id, name, email, image

I am using a next-auth prisma adapter that is automatically writing the data on OAuth signin.

when i login with OAuth the i get the error mentioned above about foreign keys with `authorId` since it will be null as an author ID hasn't been assigned to the user yet and they might not even have an author ID.

**Reply by janpio** [reply]
Are you sure it is trying to write `NULL` there? 

Random thought: Could it be that (your) MySQL only allows 1 value with `NULL` per table? Do you already have one and it would try adding the second one? Can you confirm?

**Reply by Ahmadh26** [reply]
I'm not sure about it trying to write it as NULL, but authorId is not unique so it shouldn't give an error if there's multiple author IDs that are the same

I have also tried setting a default value (eg, `-1`) by adding it directly in the modal `@default("-1")` and it still did the same error. 

How can I check if it's actually trying to set it as NULL? Shouldn't the `?` indicate that if there's no value set it as NULL?

**Reply by janpio** [reply]
The error message `Foreign key constraint failed on the field: 'authorId'` actually indicates that it is writing a value that would not follow referential integrity of where the column points to - that means a value that does not exist in `Author.id`.

You can see the actual query this way: https://pris.ly/d/logging + https://pris.ly/d/debugging

**Reply by Ahmadh26** [reply]
so here is the query being executed: 
```bash 
prisma:query SELECT 1
prisma:query BEGIN
prisma:query INSERT INTO `test`.`User` (`id`,`name`,`email`,`nsid`,`type`,`createdAt`) VALUES (?,?,?,?,?,?)
prisma:query ROLLBACK
```


here is the error 
```
PrismaClientKnownRequestError: 
Invalid `prisma.user.create()` invocation:


  Foreign key constraint failed on the field: `authorId`
    at Object.request (C:\Users\a_4n\Desktop\repos\ns-v2\node_modules\@prisma\client\runtime\index.js:45578:15)
    at async PrismaClient._request (C:\Users\a_4n\Desktop\repos\ns-v2\node_modules\@prisma\client\runtime\index.js:46405:18)
    at async handler (webpack-internal:///(api)/./pages/api/auth/register.ts:26:26)
    at async Object.apiResolver (C:\Users\a_4n\Desktop\repos\ns-v2\node_modules\next\dist\server\api-utils\node.js:182:9)
    at async DevServer.runApi (C:\Users\a_4n\Desktop\repos\ns-v2\node_modules\next\dist\server\next-server.js:386:9)
    at async Object.fn (C:\Users\a_4n\Desktop\repos\ns-v2\node_modules\next\dist\server\base-server.js:488:37)
    at async Router.execute (C:\Users\a_4n\Desktop\repos\ns-v2\node_modules\next\dist\server\router.js:228:32)
    at async DevServer.run (C:\Users\a_4n\Desktop\repos\ns-v2\node_modules\next\dist\server\base-server.js:600:29)
    at async DevServer.run (C:\Users\a_4n\Desktop\repos\ns-v2\node_modules\next\dist\server\dev\next-dev-server.js:445:20)
    at async DevServer.handleRequest (C:\Users\a_4n\Desktop\repos\ns-v2\node_modules\next\dist\server\base-server.js:307:20) {
  code: 'P2003',
  clientVersion: '3.12.0',
  meta: { field_name: 'authorId' }
}
```

I'm not sure what is happening as the value inserted into author ID is infact NULL

**Reply by janpio** [reply]
Does the same problem happen when you execute a similar SQL statement manually?
Just to double check: There is no default on the column in the database? Is there already an entry with `NULL`? If so, can you remove it and see if the script then succeeds once? (Just to make sure this is not somehow the case and happening)

**Reply by Ahmadh26** [reply]
i did a check on the database column, apparently it has the default value saved even after changing it and migrating. 
i deleted the tables and pushed (`prisma db push`) which fixed it.

**Reply by janpio** [reply]
Good.

Could you please open a bug issue for that problem though? And maybe even try to reproduce it? I assume the column once had a default, and then you removed it and migrated - but that somehow did nothing? Do you have a migration file that should have removed it?

**Reply by Ahmadh26** [reply]
i have tried to re-produce it but it didn't work. Must be a mistake i did in the process i guess? 
i have the generated migration files still, would you like me to post them here?

**Reply by janpio** [reply]
Yeah, any information can help. 

If you can not reproduce, it might be indeed that you missed something somewhere along the lines or were just unlucky. But if you _can_ reproduce, we want to confirm that and fix this sooner or later.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
